NativeAOT: ICustomMarshaller

kant2002 opened the pull request:
This is https://github.com/dotnet/runtimelab/pull/1723
Closes https://github.com/dotnet/runtimelab/issues/160

kant2002 committed:
NativeAOT: ICustomMarshaller

This is https://github.com/dotnet/runtimelab/pull/1723/files without issues.targets change

jkotas commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:
```diff
@@ -536,24 +538,20 @@ protected override void EmitMarshalFieldNativeToManaged()
 
     abstract class ByValStringMarshaller : ByValArrayMarshaller
     {
-        protected override void EmitElementCount(ILCodeStream codeStream, MarshalDirection direction)
```
This looks like a bogus change.

kant2002 replied:
indeed :(

kant2002 committed:
Remove bogus change

kant2002 committed:
Port issues.targets changes

kant2002 commented:
Given that NativeAOT lives in the runtime repo, can I somehow document in the tests difference between throwing exceptions for ICustomMarshaller ?
For example I can copy `src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs ` to `src/tests/nativeaot/SmokeTests/` and made changes at that location, or maybe I can somehow add some `#define` as part of this PR, so at least differences would be tested, documented and I can cleanup `issues.targets` even more.

MichalStrehovsky commented:
> Given that NativeAOT lives in the runtime repo, can I somehow document in the tests difference between throwing exceptions for ICustomMarshaller ?

How do the differences look like? One possible option would be to adjust the test to accept both exception types. We can ask the interop team what they think once we can see the exact diff. Some of the exceptions thrown for invalid inputs are really just a side effect of how things are implemented in the runtime and if it will be acceptable for NativeAOT to throw something else, it can be acceptable for any runtime and it's okay to adjust the test for all flavors.

kant2002 committed:
Add missing code

kant2002 committed:
Add comments with infomraiotn about different exceptions

kant2002 commented:
@MichalStrehovsky please take a look at the differences. I put them as comments, just so they would be visible. 
Some of them become quite generic like everything now is `MarshallingException`.
Others, like `typeof(System.Exception): Method '[ICustomMarshaler_TargetWindows]System.Runtime.InteropServices.Tests.ICustomMarshalerTests.MarshalerOnPointerMethod(int32*)' requires marshalling that is not yet supported by this compiler.` can be handled by ILC if it can recognize not only `Invalid` marshaller, but `NotImplemented`. In our case we clearly know that some marshalling is not supported and never will, but right now it falls in the "not yet implemeted" bucket.

kant2002 commented:
Also separate question. In order to put a block on the delegates with ref parameters I should perform some voodo. 

Right now `delegate (ref int)` => ask for marshalling of stub with this signature `{[S.P.CompilerGenerated]Internal.CompilerGenerated.<Module>.ReverseDelegateStub__TestDelegateRef(native int*)}`. Information about `ref` is lost here. It is not super important for me, but would like to make that work while I'm here.

kant2002 committed:
Tighten tests
- Remove annotations which are not needed
- Add test which specifically tests ordering for return value

kant2002 committed:
Fix all newly added tests

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/MarshalHelpers.cs:
```diff
@@ -494,6 +505,8 @@ public static partial class MarshalHelpers
                             else
                                 return MarshallerKind.ByValArray;
                         }
+                    case NativeTypeKind.CustomMarshaler:
```
These cases (below too) are now unreachable.

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:
```diff
@@ -1129,4 +1131,155 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
+            var getInstanceMethod = marshallerType.GetMethod(
+                "GetInstance",
+                new MethodSignature(MethodSignatureFlags.Static, 0, customMarshallerType, new[] { Context.GetWellKnownType(WellKnownType.String) }));
+            if (getInstanceMethod == null)
+            {
+                // ApplicationException throw here.
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
```
This looks unnecessary - there's handling for this below and in corelib.

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:
```diff
@@ -1129,4 +1131,155 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            if (!marshallerType.CanCastTo(customMarshallerType))
```
We could do this check in CoreLib and throw the correct exception - `RuntimeImports.AreTypesAssignable` is the API to use.

MichalStrehovsky commented on src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:
```diff
@@ -370,6 +380,7 @@ public static ICustomMarshaler GetInstance(string cookie)
 
         public static void Parameter_MarshalerOnValueType_ThrowsMarshalDirectiveException()
         {
+            // typeof(System.Exception): Method '[ICustomMarshaler_TargetWindows]System.Runtime.InteropServices.Tests.ICustomMarshalerTests.MarshalerOnValueTypeMethod(int32)' requires marshalling that is not yet supported by this compiler.
```
Jeremy is fixing this one and below in #63320 (we will get rid of "requires marshalling that is not yet supported" and just throw MarshalDirective).

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:
```diff
@@ -1129,4 +1131,155 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
+            }
+
+            var customMarshallerType = Context.SystemModule.GetKnownType("System.Runtime.InteropServices", "ICustomMarshaler");
+            if (customMarshallerType == null)
```
GetKnownType won't return null

jkotas replied:
This does not look addressed.

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:
```diff
@@ -1129,4 +1131,155 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
+            var marshallerType = MarshalAsDescriptor.CustomMarshallerType;
+            if (marshallerType == null)
+            {
+                // TypeLoadException throw here.
+                ThrowHelper.ThrowMarshalDirectiveException();
```
ThrowHelper.ThrowTypeLoadException?

kant2002 replied:
There no type here. Best overload is `void ThrowTypeLoadException(string @namespace, string name, string moduleName)` but I have to deduce namespace and module name. Module name I can pass to `MarshalAsDescriptor` from `EcmaSignatureParser` but isn't it overkill?

MichalStrehovsky replied:
If you change MarshalAsDescriptor to have a string instead of TypeDesc and do the conversion here, you can ask the conversion helper to throw exception instead of returning null. It will throw the right exception

kant2002 replied:
Look at this https://github.com/dotnet/runtime/pull/62921/commits/0fce5bae841a0f13632373eb258bd93f66461894 I take liberty there, but that's seems to be reasonable way. Notice: This case is where no `TypeDesc == null` and `typeName == null` as well.

MichalStrehovsky replied:
Is there anything problematic with my suggestion though ("If you change MarshalAsDescriptor to have a string instead of TypeDesc and do the conversion here, you can ask the conversion helper to throw exception")? I'm not a fan of `TypeLoadException` with no message when we instead could throw "FileLoadException: assembly X could not be found", or "TypeLoadException: Type X in assembly Y could not be found".

kant2002 replied:
This is case for `null`. Null string is still `null` and we are would not have type information. Also your suggested method was my initial attempt and this is what we settle on because there no reasonable way to create TypeDesc inside Marshaller class

kant2002 replied:
That’s the case for message “Cannot load null type”

MichalStrehovsky replied:
That field is populated as `customMarshallerType = _ecmaModule.GetTypeByCustomAttributeTypeName(managedMarshallerTypeName, false);` so this will be null for everything that didn't resolve. Not just a null string.

Maybe we can flip this to `true` and then it will never be null.

kant2002 replied:
Changes which I introduce was for this case only 
https://github.com/dotnet/runtime/blob/84926be4243143b7830e78cb7727057ed711a3cc/src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs#L387-L393

kant2002 replied:
Now I re-read and understand what you are saying.

kant2002 replied:
I did not handle just the case when type name point to separate assembly. I have to add additional test case and honestly do not want to touch that unless that's nesecssary. This seems to be fringe implementation feature. Let somebody else report issue 😄 . I would gladly help, or maybe as followup PR with tests.

MichalStrehovsky commented on src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:
```diff
@@ -386,6 +398,7 @@ public static unsafe void Parameter_MarshalerOnPointer_ThrowsMarshalDirectiveExc
         public static void Parameter_NullICustomMarshaler_ThrowsTypeLoadException()
         {
+            // typeof(System.Runtime.InteropServices.MarshalDirectiveException): Marshaling directives are invalid.
             Assert.Throws<TypeLoadException>(() => NullCustomMarshalerMethod(""));
```
I think we just have throw the right exception from the compiler (I commented on the spot).

MichalStrehovsky commented on src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:
```diff
@@ -394,6 +407,7 @@ public static void Parameter_NullICustomMarshaler_ThrowsTypeLoadException()
 
         public static void Parameter_NotICustomMarshaler_ThrowsApplicationException()
         {
+            // typeof(System.Runtime.InteropServices.MarshalDirectiveException): Marshaling directives are invalid.
```
I commented on the spot that I think is responsible for this being different.

MichalStrehovsky reviewed (commented):
I think we'll be able to greatly reduce (maybe even eliminate) the number of exception differences.

kant2002 committed:
Remove no longer needed cases

kant2002 committed:
Handle more cases

kant2002 committed:
Remove one incompatibility

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/InteropHelpers.cs:
```diff
@@ -591,6 +593,32 @@ internal static unsafe void CleanupVariant(IntPtr pDstNativeVariant)
 #endif
         }
 
+        public static unsafe object InitializeCustomMarshaller(RuntimeTypeHandle pParameterType, RuntimeTypeHandle pMarshallerType, RuntimeTypeHandle pICustomMarshallerType, string cookie, delegate*<string, object> getInstanceMethod)
```
`typeof(ICustomMarshaller)` or `EETypePtrOf<ICustomMarshaller>` would work just fine, no need to pass it from the generated IL.

kant2002 replied:
Excellent. I was scary to use that because you told me that Type creation is relatively expensive. So I thought this is the case too.

kant2002 committed:
Apply feedback from PR

kant2002 committed:
This is allow throw `new TypeLoadException()` which I need when marshaller type is null. There no type information to pass.

kant2002 committed:
Uncomment code, this was debug mistake

kant2002 committed:
Add test for non-existing type

kant2002 committed:
Make TypeLoadException more consistent with CoreCLR

kant2002 commented:
@MichalStrehovsky what's the plan how to move forward?

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:
```diff
@@ -1129,4 +1131,146 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
+            var getInstanceMethod = marshallerType.GetMethod(
+                "GetInstance",
+                new MethodSignature(MethodSignatureFlags.Static, 0, customMarshallerType, new[] { Context.GetWellKnownType(WellKnownType.String) }));
+            if (ManagedType.IsValueType || ManagedType.IsPointer)
```
```suggestion
            if (ManagedType.IsValueType || ManagedType.IsPointer || ManagedType.IsFunctionPointer)
```

MichalStrehovsky reviewed (commented):
> @MichalStrehovsky what's the plan how to move forward?

I think it's very close, thanks for your patience! Could you also addressing the old feedback?

https://github.com/dotnet/runtime/pull/62921#discussion_r785599675 - the call to GetTypeByCustomAttributeTypeName should pass `true`. Then it will throw an exception if the type cannot be found (including handling the case when the assembly doesn't exist) and the type returned from it will never be null.

https://github.com/dotnet/runtime/pull/62921#discussion_r781703380

Once Jeremy's change gets in, we'll be able to re-test how close we're with passing the tests. So far we didn't have to do unnatural things to keep exceptional behavior the same, which is basically my bar for exception compatibility.

kant2002 commented:
> the call to GetTypeByCustomAttributeTypeName should pass true

I do understand that idea. I'm not convinced that this is proper fix. Maybe I miss something. Let me explain.
If I add `true` to `GetTypeByCustomAttributeTypeName` then exception would be thrown at compiler phase and I **assume** that it will crash compiler. In itself this is reasonable behavior, but this is change from CoreCLR. I would like to confirm that my assumption is correct, and ILC crash is correct inference.

Right now I handle case of incorrect type, and I even add test for that. Way how I do that seems to be fine and reasonable, so maybe I should not change what's I have right now. *If* ILC would crash ,then I should undo that test.

> Once Jeremy's change gets in

Ok. I will wait. 

So far, I only want to clarify `GetTypeByCustomAttributeTypeName` course of action and will monitor sibling PR.

kant2002 committed:
Update src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs

Co-authored-by: Michal Strehovský <MichalStrehovsky@users.noreply.github.com>

MichalStrehovsky commented:
Can you help me understand that concern? In your change you have:

```csharp
            if (marshallerType == null || marshallerType.IsGenericDefinition)
            {
                ThrowHelper.ThrowTypeLoadException(marshallerTypeName ?? string.Empty, module);
            }
```

That's an exception throw. Those are the same kinds of throws that will happen if you pass true to GetTypeByCustomAttributeTypeName. E.g.

https://github.com/dotnet/runtime/blob/e4bfedd847990795eb28cda4008b5ff4c58fd561/src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs#L255-L256

These are all caught at the same spot.

kant2002 committed:
Use well-known API for checking for type name correctness

kant2002 committed:
Document changes in behaviour

kant2002 committed:
Uncomment test

kant2002 commented:
@MichalStrehovsky so far still have 2 differences in same places, they are shown in tests. I test by merging Jeremy's changes from latest main.

kant2002 commented:
The 2b44656631eb27d71531ec366d3f1999372d42d9 irrelevant. Both variants produce same unhandled exception

**Managed exception**
```
Unhandled Exception: System.Runtime.InteropServices.MarshalDirectiveException: Marshaling directives are invalid.
   at Internal.Runtime.TypeLoaderExceptionHelper.CreateMarshalDirectiveException(ExceptionStringID) + 0x4a
   at Internal.Runtime.CompilerHelpers.ThrowHelpers.ThrowMarshalDirectiveException(ExceptionStringID) + 0x1c
   at ICustomMarshaler_TargetUnix!<BaseAddress>+0x5f6a7b
```

**Unmanaged exception**
```
> kc
 # Call Site
00 ntdll!NtRaiseException
01 KERNELBASE!RaiseFailFastException
02 ICustomMarshaler_TargetWindows!S_P_CoreLib_Interop_Kernel32__RaiseFailFastException_0
03 ICustomMarshaler_TargetWindows!S_P_CoreLib_Interop_Kernel32__RaiseFailFastException
04 ICustomMarshaler_TargetWindows!S_P_CoreLib_System_RuntimeExceptionHelpers__FailFast_1
05 ICustomMarshaler_TargetWindows!S_P_CoreLib_System_RuntimeExceptionHelpers__RuntimeFailFast
06 ICustomMarshaler_TargetWindows!S_P_CoreLib_System_Runtime_EH__UnhandledExceptionFailFastViaClasslib
07 ICustomMarshaler_TargetWindows!S_P_CoreLib_System_Runtime_EH__DispatchEx
08 ICustomMarshaler_TargetWindows!S_P_CoreLib_System_Runtime_EH__RhThrowEx
09 ICustomMarshaler_TargetWindows!RhpThrowEx
0a ICustomMarshaler_TargetWindows!S_P_CoreLib_Internal_Runtime_TypeLoaderExceptionHelper__CreateMarshalDirectiveException
0b ICustomMarshaler_TargetWindows!S_P_CoreLib_Internal_Runtime_CompilerHelpers_ThrowHelpers__ThrowMarshalDirectiveException
0c ICustomMarshaler_TargetWindows!Internal_CompilerGenerated__Module___<ReverseDelegateStub>ICustomMarshaler_TargetWindows_System_Runtime_InteropServices_Tests_ICustomMarshalerTests_TestDelegateRef
0d CustomMarshalersPrimitives!NativeParseIntDelegateRef
0e ICustomMarshaler_TargetWindows!ICustomMarshaler_TargetWindows_System_Runtime_InteropServices_Tests_ICustomMarshalerTests__CustomMarshallerWithDelegateRef
0f ICustomMarshaler_TargetWindows!ICustomMarshaler_TargetWindows_System_Runtime_InteropServices_Tests_ICustomMarshalerTests___c___DelegateParameter_MarshalerOnRefInt_ThrowsMarshalDirectiveException_b__83_0
10 ICustomMarshaler_TargetWindows!xunit_assert_Xunit_Assert__RecordException_0
11 ICustomMarshaler_TargetWindows!xunit_assert_Xunit_Assert__Throws_0<System___Canon>
12 ICustomMarshaler_TargetWindows!ICustomMarshaler_TargetWindows_System_Runtime_InteropServices_Tests_ICustomMarshalerTests__DelegateParameter_MarshalerOnRefInt_ThrowsMarshalDirectiveException
13 ICustomMarshaler_TargetWindows!ICustomMarshaler_TargetWindows_System_Runtime_InteropServices_Tests_ICustomMarshalerTests__Main
14 ICustomMarshaler_TargetWindows!ICustomMarshaler_TargetWindows__Module___MainMethodWrapper
15 ICustomMarshaler_TargetWindows!ICustomMarshaler_TargetWindows__Module___StartupCodeMain
16 ICustomMarshaler_TargetWindows!wmain
17 ICustomMarshaler_TargetWindows!invoke_main
18 ICustomMarshaler_TargetWindows!__scrt_common_main_seh
19 ICustomMarshaler_TargetWindows!__scrt_common_main
1a ICustomMarshaler_TargetWindows!wmainCRTStartup
1b KERNEL32!BaseThreadInitThunk
1c ntdll!RtlUserThreadStart
```

kant2002 commented:
Add test for that specific scenario https://github.com/dotnet/runtime/pull/64137

MichalStrehovsky commented on src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:
```diff
@@ -658,15 +677,15 @@ private class InnerCustomMarshaler : ILargeInterface, ICustomMarshaler
 
         public static void DelegateParameter_MarshalerOnRefInt_ThrowsMarshalDirectiveException()
         {
+            // throw MarshalDirectiveException, but it is unhandled (?)
```
We should probably fix this so that GetFunctionPointerForDelegate doesn't return "broken" function pointers that are just going to throw. Should be a separate pull request. Let's not enable this test for NativeAOT for now.

kant2002 replied:
Do this problem have tracking issue? I would like to change issue which prevent test from running to something different to indicate that https://github.com/dotnet/runtimelab/issues/160 indeed closed.

kant2002 replied:
Can I hide that test under 
```
if (RuntimeInformation.OSPlatform == Platform.Windows && !RuntimeFeatured.IsDynamicCodeSupported)
```

that way I will capture your words in the test. Maybe just 
```
if (!RuntimeFeatured.IsDynamicCodeSupported)
```

kant2002 replied:
I think I rushing again. I agree with Michal, `GetFunctionPointerForDelegate` should throw if somebody pass delegate which contains broken marshaling patterns. That would automatically fix current issue because throw happens before delegate was passed to native code.

My idea to create(if needed) separate issue, and re-assign test to that issue, because right now all related tests for  ICustomMarshaler passing. This problem in my understanding related to any delegates marshaling which contains unsupported configuration.

jkotas replied:
It would be best to use `src\libraries\Common\tests\TestUtilities\System\PlatformDetection.cs` for this.

jkotas replied:
> I agree with Michal, GetFunctionPointerForDelegate should throw if somebody pass delegate which contains broken marshaling patterns.

FWIW, it is not how CoreCLR works. CoreCLR will throw only once you actually try to call the delegate.

kant2002 replied:
I cannot add `src\libraries\Common\tests\TestUtilities\System\PlatformDetection.cs` alone because it complains about `IsWindows` properties. If I add `src\libraries\Common\tests\TestUtilities\System\PlatformDetection.Windows.cs` test build complain about `GenerateDllImportAttribute`. If I add reference to `TestUtilities.csproj` then `src\tests\build.sh` complain about not restored assets for the project.

I look for usages of this project in the tests, but seems to be it is used only in libraries tests. Please advice.

jkotas replied:
If it is hard to use the full implementation from libraries currently, I would create a simplified local implementation of  PlatformDetection for runtime tests, with the subset of properties, and use that instead.

@jkoritzinsky Do you happen to have anything better in the plan for the runtime test consolidation?

jkoritzinsky replied:
I think for now, making a copy either locally or in the `CoreCLRTestLibrary` assembly is the way to go. As we move tests over, we might decide to consolidate some of these utility libraries.

MichalStrehovsky commented on src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:
```diff
@@ -596,6 +614,7 @@ public class ThrowingCleanUpNativeDataCustomMarshaler : ICustomMarshaler
 
         public static void Field_ParentIsStruct_ThrowsTypeLoadException()
         {
+            //  typeof(System.Runtime.InteropServices.MarshalDirectiveException): Struct 'StructWithCustomMarshalerField' requires marshalling that is not yet supported by this compiler.
```
I think it should be fine to update this test to also accept MarshalDirectiveException. TypeLoadException is odd.

kant2002 replied:
I change this exception type in this PR. that was fairly trivial. I can move that change to separate PR if needed.

MichalStrehovsky replied:
Definitely a separate pull request. I somewhat doubt changing the exception thrown out of CoreCLR will be accepted, but one can always try.

kant2002 replied:
Remove code. Somebody has to rerun pipeline

jkoritzinsky replied:
Is this comment still necessary?

MichalStrehovsky reviewed (commented):
Looks good otherwise.

@jkoritzinsky could you please have a look at the changes to the tests?

kant2002 committed:
Do not enable Interop/ICustomMarshaler/Primitives/ICustomMarshaler_TargetWindows

kant2002 commented:
/azp run

kant2002 committed:
Throw TypeLoadException on when apply ICustomMarshaller to field

jkotas commented on src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -280,8 +280,8 @@
     <Compile Include="$(BclSourcesRoot)\System\Runtime\InteropServices\CustomMarshalers\EnumeratorViewOfEnumVariant.cs" />
-    <Compile Include="$(BclSourcesRoot)\System\Runtime\InteropServices\CustomMarshalers\ExpandoToDispatchExMarshaler.cs" />
-    <Compile Include="$(BclSourcesRoot)\System\Runtime\InteropServices\CustomMarshalers\TypeToTypeInfoMarshaler.cs" />
+    <Compile Include="$(CommonPath)\System\Runtime\InteropServices\CustomMarshalers\ExpandoToDispatchExMarshaler.cs" />
```
I do not think we want to add these to NativeAOT just to make some random tests pass. 

These are part of built-in COM interop support. We do not have built-in support in NativeAOT. It does not make sense to add a small incoherent bits of it.

kant2002 replied:
So I should undo changes and mark test as CoreCLR specific?

jkotas replied:
It would be best to mark it as built-in COM interop specific using `PlatformDetection.SupportsComInterop`.

kant2002 replied:
Thank you.

kant2002 committed:
Do not run unsupported case inside test

kant2002 committed:
Mark tests according to the passing status

kant2002 committed:
Add proper feature check

jkoritzinsky commented on src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:
```diff
@@ -699,7 +718,11 @@ public static int Main(String[] args)
                 Field_ParentIsStruct_ThrowsTypeLoadException();
                 Parameter_DifferentCustomMarshalerType_MarshalsCorrectly();
-                DelegateParameter_MarshalerOnRefInt_ThrowsMarshalDirectiveException();
+                if (SupportsComInterop)
```
This feels like the wrong name for this condition. Maybe `SupportsExceptionInterop` would be a better name?

jkoritzinsky reviewed (approved):
LGTM with a few nits

kant2002 committed:
Rename SupportsComInterop to SupportsExceptionInterop

kant2002 commented:
I have enough endorphin supplies for today, but maybe this PR can also be merged? I seems to be resolve all issues from interested parties.

jkotas commented:
> I seems to be resolve all issues from interested parties.

There two small comments do not appear to be addressed:
- https://github.com/dotnet/runtime/pull/62921#discussion_r781703380
- https://github.com/dotnet/runtime/pull/62921#discussion_r796082492

kant2002 committed:
Fix PR feedback

jkotas reviewed (approved):
LGTM. Thank you!
